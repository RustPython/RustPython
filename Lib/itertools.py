import operator

class accumulate:
    def __init__(self, iterable, func = operator.add):
        self.it = iter(iterable)
        self._total = None
        self.func = func
        
    def __iter__(self):
        return self

    def __next__(self):
        if not self._total:
            self._total = next(self.it)
            return self._total
        else:
            element = next(self.it)
            try:
                self._total = self.func(self._total, element)
            except:
                raise TypeError("unsupported operand type")
            return self._total
                
## Adapted from:
## https://bitbucket.org/pypy/pypy/src/c1aa74c06e86/lib_pypy/itertools.py#cl-34
class chain:
    def __init__(self, *iterables):
        self._iterables_iter = iter(map(iter, iterables))
        # little trick for the first chain.__next__() call
        self._cur_iterable_iter = iter([])

    def __iter__(self):
        return self
    
    def __next__(self):
        while True:
            try:
                return next(self._cur_iterable_iter)
            except StopIteration:
                self._cur_iterable_iter = next(self._iterables_iter)
    
    @classmethod
    def from_iterable(cls, iterable):
        for it in iterable:
            for element in it:
                yield element
                
class combinations:
    def __init__(self, iterable, r):
        self.pool = tuple(iterable)
        self.n = len(self.pool)
        self.r = r
        self.indices = list(range(self.r))
        self.zero = False

    def __iter__(self):
        return self

    def __next__(self):
        if self.r > self.n:
            raise StopIteration
        if not self.zero:
            self.zero = True
            return tuple(self.pool[i] for i in self.indices)
        else:
            try:
                for i in reversed(range(self.r)):
                    if self.indices[i] != i + self.n - self.r:
                        break
                self.indices[i] += 1
                for j in range(i+1, self.r):
                    self.indices[j] = self.indices[j-1] + 1
                return tuple(self.pool[i] for i in self.indices)
            except:
                raise StopIteration
                
class combinations_with_replacement:
    def __init__(self, iterable, r):
        self.pool = tuple(iterable)
        self.n = len(self.pool)
        self.r = r
        self.indices = [0] * self.r
        self.zero = False
        
    def __iter__(self):
        return self
    
    def __next__(self):
        if not self.n and self.r:
            raise StopIteration
        if not self.zero:
            self.zero = True
            return tuple(self.pool[i] for i in self.indices)
        else:
            try:
                for i in reversed(range(self.r)):
                    if self.indices[i] != self.n - 1:
                        break
                self.indices[i:] = [self.indices[i] + 1] * (self.r - i)
                return tuple(self.pool[i] for i in self.indices)
            except:
                raise StopIteration
                
## Literally copied from
##https://bitbucket.org/pypy/pypy/src/c1aa74c06e86/lib_pypy/itertools.py#cl-63 
class compress:
    def __init__(self, data, selectors):
        self.data = iter(data)
        self.selectors = iter(selectors)

    def __iter__(self):
        return self

    def __next__(self):
        while True:
            next_item = next(self.data)
            next_selector = next(self.selectors)
            if bool(next_selector):
                return next_item
                
## Adapted from
##https://bitbucket.org/pypy/pypy/src/c1aa74c06e86/lib_pypy/itertools.py#cl-79
## I mimicked the > python3.1 behavior
class count:
    """
    Input is an int or a float. The original Python 3 implementation
    includes also complex numbers... but it still is not implemented 
    in Brython as complex type is NotImplemented
    """
    def __init__(self, start = 0, step = 1):
        if not isinstance(start, (int, float)):
            raise TypeError('a number is required')
        self.times = start - step
        self.step = step

    def __iter__(self):
        return self

    def __next__(self):
        self.times += self.step
        return self.times

    def __repr__(self):
        return 'count(%d)' % (self.times + self.step)

## Literally copied from
##https://bitbucket.org/pypy/pypy/src/c1aa74c06e86/lib_pypy/itertools.py#cl-112        
class cycle:
    def __init__(self, iterable):
        self._cur_iter = iter(iterable)
        self._saved = []
        self._must_save = True
        
    def __iter__(self):
        return self

    def __next__(self):
        try:
            next_elt = next(self._cur_iter)
            if self._must_save:
                self._saved.append(next_elt)
        except StopIteration:
            self._cur_iter = iter(self._saved)
            next_elt = next(self._cur_iter)
            self._must_save = False
        return next_elt
        
## Literally copied from
##https://bitbucket.org/pypy/pypy/src/c1aa74c06e86/lib_pypy/itertools.py#cl-149
class dropwhile:
    def __init__(self, predicate, iterable):
        self._predicate = predicate
        self._iter = iter(iterable)
        self._dropped = False

    def __iter__(self):
        return self

    def __next__(self):
        value = next(self._iter)
        if self._dropped:
            return value
        while self._predicate(value):
            value = next(self._iter)
        self._dropped = True
        return value
        
## Adapted from
##https://bitbucket.org/pypy/pypy/src/c1aa74c06e86/lib_pypy/itertools.py#cl-261
class filterfalse:
    def __init__(self, predicate, iterable):
        # Make sure iterable *IS* iterable
        self._iter = iter(iterable)
        if predicate is None:
            self._predicate = bool
        else:
            self._predicate = predicate

    def __iter__(self):
        return self
    def __next__(self):
        next_elt = next(self._iter)
        while True:
            if not self._predicate(next_elt):
                return next_elt
            next_elt = next(self._iter)

class groupby:
    # [k for k, g in groupby('AAAABBBCCDAABBB')] --> A B C D A B
    # [list(g) for k, g in groupby('AAAABBBCCD')] --> AAAA BBB CC D
    def __init__(self, iterable, key=None):
        if key is None:
            key = lambda x: x
        self.keyfunc = key
        self.it = iter(iterable)
        self.tgtkey = self.currkey = self.currvalue = object()
    def __iter__(self):
        return self
    def __next__(self):
        while self.currkey == self.tgtkey:
            self.currvalue = next(self.it)    # Exit on StopIteration
            self.currkey = self.keyfunc(self.currvalue)
        self.tgtkey = self.currkey
        return (self.currkey, self._grouper(self.tgtkey))
    def _grouper(self, tgtkey):
        while self.currkey == tgtkey:
            yield self.currvalue
            self.currvalue = next(self.it)    # Exit on StopIteration
            self.currkey = self.keyfunc(self.currvalue)
          
## adapted from
##https://bitbucket.org/pypy/pypy/src/c1aa74c06e86/lib_pypy/itertools.py#cl-323
class islice:
    def __init__(self, iterable, *args):
        s = slice(*args)
        self.start, self.stop, self.step = s.start or 0, s.stop, s.step
        if not isinstance(self.start, int):
           raise ValueError("Start argument must be an integer")
        if self.stop != None and not isinstance(self.stop, int):
            raise ValueError("Stop argument must be an integer or None")
        if self.step is None:
            self.step = 1
        if self.start<0 or (self.stop != None and self.stop<0
           ) or self.step<=0:
            raise ValueError("indices for islice() must be positive")
        self.it = iter(iterable)
        self.donext = None
        self.cnt = 0

    def __iter__(self):
        return self

    def __next__(self):
        nextindex = self.start
        if self.stop != None and nextindex >= self.stop:
            raise StopIteration
        while self.cnt <= nextindex:
            nextitem = next(self.it)
            self.cnt += 1
        self.start += self.step 
        return nextitem
        
class permutations:
    def __init__(self, iterable, r = None):
        self.pool = tuple(iterable)
        self.n = len(self.pool)
        self.r = self.n if r is None else r
        self.indices = list(range(self.n))
        self.cycles = list(range(self.n, self.n - self.r, -1))
        self.zero = False
        self.stop = False

    def __iter__(self):
        return self

    def __next__(self):
        indices = self.indices
        if self.r > self.n:
            raise StopIteration
        if not self.zero:
            self.zero = True
            return tuple(self.pool[i] for i in indices[:self.r])
        
        i = self.r - 1
        while i >= 0:
            j = self.cycles[i] - 1
            if j > 0:
                self.cycles[i] = j
                indices[i], indices[-j] = indices[-j], indices[i]
                return tuple(self.pool[i] for i in indices[:self.r])
            self.cycles[i] = len(indices) - i
            n1 = len(indices) - 1
            assert n1 >= 0
            num = indices[i]
            for k in range(i, n1):
                indices[k] = indices[k+1]
            indices[n1] = num
            i -= 1
        raise StopIteration

# copied from Python documentation on itertools.product
def product(*args, repeat=1):
    # product('ABCD', 'xy') --> Ax Ay Bx By Cx Cy Dx Dy
    # product(range(2), repeat=3) --> 000 001 010 011 100 101 110 111
    pools = [tuple(pool) for pool in args] * repeat
    result = [[]]
    for pool in pools:
        result = [x+[y] for x in result for y in pool]
    for prod in result:
        yield tuple(prod)


## adapted from
##https://bitbucket.org/pypy/pypy/src/c1aa74c06e86/lib_pypy/itertools.py#cl-392        

## (Brython) 
## renamed to _product : the implementation fails for product('abc', [])
## with CPython 3.x
class _product:
    def __init__(self, *args, **kw):
        if len(kw) > 1:
            raise TypeError("product() takes at most 1 argument (%d given)" %
                             len(kw))
        self.repeat = kw.get('repeat', 1)
        if not isinstance(self.repeat, int):
            raise TypeError("integer argument expected, got %s" %
                             type(self.repeat))
        self.gears = [x for x in args] * self.repeat
        self.num_gears = len(self.gears)
        # initialization of indicies to loop over
        self.indicies = [(0, len(self.gears[x]))
                         for x in range(0, self.num_gears)]
        self.cont = True
        self.zero = False

    def roll_gears(self):
        # Starting from the end of the gear indicies work to the front
        # incrementing the gear until the limit is reached. When the limit
        # is reached carry operation to the next gear
        should_carry = True
        for n in range(0, self.num_gears):
            nth_gear = self.num_gears - n - 1
            if should_carry:
                count, lim = self.indicies[nth_gear]
                count += 1
                if count == lim and nth_gear == 0:
                    self.cont = False
                if count == lim:
                    should_carry = True
                    count = 0
                else:
                    should_carry = False
                self.indicies[nth_gear] = (count, lim)  
            else:
                break

    def __iter__(self):
        return self

    def __next__(self):
        if self.zero:
            raise StopIteration
        if self.repeat > 0:
            if not self.cont:
                raise StopIteration
            l = []
            for x in range(0, self.num_gears):
                index, limit = self.indicies[x]
                print('itertools 353',self.gears,x,index)
                l.append(self.gears[x][index])
            self.roll_gears()
            return tuple(l)
        elif self.repeat == 0:
            self.zero = True
            return ()
        else:
            raise ValueError("repeat argument cannot be negative")
            
## Literally copied from
##https://bitbucket.org/pypy/pypy/src/c1aa74c06e86/lib_pypy/itertools.py#cl-441
class repeat:
    def __init__(self, obj, times=None):
        self._obj = obj
        if times is not None:
            range(times) # Raise a TypeError
            if times < 0:
                times = 0
        self._times = times
        
    def __iter__(self):
        return self

    def __next__(self):
        # __next__() *need* to decrement self._times when consumed
        if self._times is not None:
            if self._times <= 0: 
                raise StopIteration()
            self._times -= 1
        return self._obj

    def __repr__(self):
        if self._times is not None:
            return 'repeat(%r, %r)' % (self._obj, self._times)
        else:
            return 'repeat(%r)' % (self._obj,)

    def __len__(self):
        if self._times == -1 or self._times is None:
            raise TypeError("len() of uniszed object")
        return self._times

## Adapted from
##https://bitbucket.org/pypy/pypy/src/c1aa74c06e86/lib_pypy/itertools.py#cl-489
class starmap(object):
    def __init__(self, function, iterable):
        self._func = function
        self._iter = iter(iterable)

    def __iter__(self):
        return self

    def __next__(self):
        t = next(self._iter)
        return self._func(*t)
        
## Literally copied from
##https://bitbucket.org/pypy/pypy/src/c1aa74c06e86/lib_pypy/itertools.py#cl-520
class takewhile(object):
    def __init__(self, predicate, iterable):
        self._predicate = predicate
        self._iter = iter(iterable)

    def __iter__(self):
        return self

    def __next__(self):
        value = next(self._iter)
        if not self._predicate(value):
            raise StopIteration()
        return value

## Almost literal from
##https://bitbucket.org/pypy/pypy/src/c1aa74c06e86/lib_pypy/itertools.py#cl-547
class TeeData(object):
    def __init__(self, iterator):
        self.data = []
        self._iter = iterator

    def __getitem__(self, i):
        # iterates until 'i' if not done yet
        while i>= len(self.data):
            self.data.append(next(self._iter))
        return self.data[i]


class TeeObject(object):
    def __init__(self, iterable=None, tee_data=None):
        if tee_data:
            self.tee_data = tee_data
            self.pos = 0
        # <=> Copy constructor
        elif isinstance(iterable, TeeObject):
            self.tee_data = iterable.tee_data
            self.pos = iterable.pos
        else:
            self.tee_data = TeeData(iter(iterable))
            self.pos = 0
            
    def __next__(self):
        data = self.tee_data[self.pos]
        self.pos += 1
        return data
    
    def __iter__(self):
        return self


def tee(iterable, n=2):
    if isinstance(iterable, TeeObject):
        return tuple([iterable] +
        [TeeObject(tee_data=iterable.tee_data) for i in range(n - 1)])
    tee_data = TeeData(iter(iterable))
    return tuple([TeeObject(tee_data=tee_data) for i in range(n)])

class zip_longest:
    def __init__(self, *args, fillvalue = None):
        self.args = [iter(arg) for arg in args]
        self.fillvalue = fillvalue
        self.units = len(args)
    
    def __iter__(self):
        return self
    
    def __next__(self):
        temp = []
        nb = 0
        for i in range(self.units):
            try:
                temp.append(next(self.args[i]))
                nb += 1
            except StopIteration:
                temp.append(self.fillvalue)
        if nb==0:
            raise StopIteration
        return tuple(temp)
