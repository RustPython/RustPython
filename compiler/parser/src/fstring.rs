use self::FStringErrorType::*;
use crate::{
    ast::{Constant, ConversionFlag, Expr, ExprKind, Location},
    error::{FStringError, FStringErrorType, ParseError},
    parser::parse_expression,
};
use itertools::Itertools;
use std::{iter, mem, str};

struct FStringParser<'a> {
    chars: iter::Peekable<str::Chars<'a>>,
    str_location: Location,
    recurse_lvl: u8,
}

impl<'a> FStringParser<'a> {
    fn new(source: &'a str, str_location: Location, recurse_lvl: u8) -> Self {
        Self {
            chars: source.chars().peekable(),
            str_location,
            recurse_lvl,
        }
    }

    #[inline]
    fn expr(&self, node: ExprKind) -> Expr {
        Expr::new(self.str_location, node)
    }

    fn parse_formatted_value(&mut self) -> Result<Vec<Expr>, FStringErrorType> {
        let mut expression = String::new();
        let mut spec = None;
        let mut delims = Vec::new();
        let mut conversion = ConversionFlag::None;
        let mut self_documenting = false;
        let mut trailing_seq = String::new();

        while let Some(ch) = self.chars.next() {
            match ch {
                // can be integrated better with the remainign code, but as a starting point ok
                // in general I would do here a tokenizing of the fstrings to omit this peeking.
                '!' if self.chars.peek() == Some(&'=') => {
                    expression.push_str("!=");
                    self.chars.next();
                }

                '=' if self.chars.peek() == Some(&'=') => {
                    expression.push_str("==");
                    self.chars.next();
                }

                '>' if self.chars.peek() == Some(&'=') => {
                    expression.push_str(">=");
                    self.chars.next();
                }

                '<' if self.chars.peek() == Some(&'=') => {
                    expression.push_str("<=");
                    self.chars.next();
                }

                '!' if delims.is_empty() && self.chars.peek() != Some(&'=') => {
                    if expression.trim().is_empty() {
                        return Err(EmptyExpression);
                    }

                    conversion = match self.chars.next() {
                        Some('s') => ConversionFlag::Str,
                        Some('a') => ConversionFlag::Ascii,
                        Some('r') => ConversionFlag::Repr,
                        Some(_) => {
                            return Err(InvalidConversionFlag);
                        }
                        None => {
                            return Err(ExpectedRbrace);
                        }
                    };

                    if let Some(&peek) = self.chars.peek() {
                        if peek != '}' && peek != ':' {
                            return Err(ExpectedRbrace);
                        }
                    } else {
                        return Err(ExpectedRbrace);
                    }
                }

                // match a python 3.8 self documenting expression
                // format '{' PYTHON_EXPRESSION '=' FORMAT_SPECIFIER? '}'
                '=' if self.chars.peek() != Some(&'=') && delims.is_empty() => {
                    self_documenting = true;
                }

                ':' if delims.is_empty() => {
                    let mut nested = 0;
                    let mut spec_constructor = Vec::new();
                    let mut constant_piece = String::new();
                    let mut formatted_value_piece = String::new();
                    while let Some(&next) = self.chars.peek() {
                        match next {
                            '{' if nested > 0 => {
                                nested += 1;
                                formatted_value_piece.push(next);
                            }
                            '}' if nested > 0 => {
                                nested -= 1;
                                if nested == 0 {
                                    formatted_value_piece.push(next);
                                    let values = FStringParser::new(
                                        &formatted_value_piece,
                                        Location::default(),
                                        &self.recurse_lvl + 1,
                                    )
                                    .parse()?;
                                    spec_constructor.push(values
                                        .into_iter()
                                        .exactly_one()
                                        .expect("Expected formatted value to produce exactly one expression.")
                                    );
                                    formatted_value_piece.clear();
                                } else {
                                    formatted_value_piece.push(next);
                                }
                            }
                            _ if nested > 0 => {
                                formatted_value_piece.push(next);
                            }
                            '{' => {
                                nested += 1;
                                if !constant_piece.is_empty() {
                                    spec_constructor.push(self.expr(ExprKind::Constant {
                                        value: constant_piece.to_owned().into(),
                                        kind: None,
                                    }));
                                    constant_piece.clear();
                                }
                                formatted_value_piece.push(next);
                                formatted_value_piece.push(' ');
                            }
                            '}' => break,
                            _ => {
                                constant_piece.push(next);
                            }
                        }
                        self.chars.next();
                    }
                    if !constant_piece.is_empty() {
                        spec_constructor.push(self.expr(ExprKind::Constant {
                            value: constant_piece.to_owned().into(),
                            kind: None,
                        }));
                        constant_piece.clear();
                    }
                    if nested > 0 {
                        return Err(UnclosedLbrace);
                    }
                    spec = Some(Box::new(self.expr(ExprKind::JoinedStr {
                        values: spec_constructor,
                    })))
                }
                '(' | '{' | '[' => {
                    expression.push(ch);
                    delims.push(ch);
                }
                ')' => {
                    if delims.pop() != Some('(') {
                        return Err(MismatchedDelimiter);
                    }
                    expression.push(ch);
                }
                ']' => {
                    if delims.pop() != Some('[') {
                        return Err(MismatchedDelimiter);
                    }
                    expression.push(ch);
                }
                '}' if !delims.is_empty() => {
                    if delims.pop() != Some('{') {
                        return Err(MismatchedDelimiter);
                    }
                    expression.push(ch);
                }
                '}' => {
                    if expression.is_empty() {
                        return Err(EmptyExpression);
                    }
                    let ret = if !self_documenting {
                        vec![self.expr(ExprKind::FormattedValue {
                            value: Box::new(
                                parse_fstring_expr(&expression)
                                    .map_err(|e| InvalidExpression(Box::new(e.error)))?,
                            ),
                            conversion: conversion as _,
                            format_spec: spec,
                        })]
                    } else {
                        vec![
                            self.expr(ExprKind::Constant {
                                value: Constant::Str(expression.clone() + "="),
                                kind: None,
                            }),
                            self.expr(ExprKind::Constant {
                                value: trailing_seq.into(),
                                kind: None,
                            }),
                            self.expr(ExprKind::FormattedValue {
                                value: Box::new(
                                    parse_fstring_expr(&expression)
                                        .map_err(|e| InvalidExpression(Box::new(e.error)))?,
                                ),
                                conversion: (if conversion == ConversionFlag::None && spec.is_none()
                                {
                                    ConversionFlag::Repr
                                } else {
                                    conversion
                                }) as _,
                                format_spec: spec,
                            }),
                        ]
                    };
                    return Ok(ret);
                }
                '"' | '\'' => {
                    expression.push(ch);
                    for next in &mut self.chars {
                        expression.push(next);
                        if next == ch {
                            break;
                        }
                    }
                }
                ' ' if self_documenting => {
                    trailing_seq.push(ch);
                }
                _ => {
                    if self_documenting {
                        return Err(ExpectedRbrace);
                    }
                    expression.push(ch);
                }
            }
        }
        Err(UnclosedLbrace)
    }

    fn parse(mut self) -> Result<Vec<Expr>, FStringErrorType> {
        if self.recurse_lvl >= 2 {
            return Err(ExpressionNestedTooDeeply);
        }

        let mut content = String::new();
        let mut values = vec![];

        while let Some(ch) = self.chars.next() {
            match ch {
                '{' => {
                    if let Some('{') = self.chars.peek() {
                        self.chars.next();
                        content.push('{');
                    } else {
                        if !content.is_empty() {
                            values.push(self.expr(ExprKind::Constant {
                                value: mem::take(&mut content).into(),
                                kind: None,
                            }));
                        }

                        values.extend(self.parse_formatted_value()?);
                    }
                }
                '}' => {
                    if let Some('}') = self.chars.peek() {
                        self.chars.next();
                        content.push('}');
                    } else {
                        return Err(UnopenedRbrace);
                    }
                }
                _ => {
                    content.push(ch);
                }
            }
        }

        if !content.is_empty() {
            values.push(self.expr(ExprKind::Constant {
                value: content.into(),
                kind: None,
            }))
        }

        Ok(values)
    }
}

fn parse_fstring_expr(source: &str) -> Result<Expr, ParseError> {
    let fstring_body = format!("({})", source);
    parse_expression(&fstring_body, "<fstring>")
}

/// Parse an fstring from a string, located at a certain position in the sourcecode.
/// In case of errors, we will get the location and the error returned.
pub fn parse_located_fstring(source: &str, location: Location) -> Result<Vec<Expr>, FStringError> {
    FStringParser::new(source, location, 0)
        .parse()
        .map_err(|error| FStringError { error, location })
}

#[cfg(test)]
mod tests {
    use super::*;

    fn parse_fstring(source: &str) -> Result<Vec<Expr>, FStringErrorType> {
        FStringParser::new(source, Location::default(), 0).parse()
    }

    #[test]
    fn test_parse_fstring() {
        let source = "{a}{ b }{{foo}}";
        let parse_ast = parse_fstring(source).unwrap();

        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_parse_fstring_nested_spec() {
        let source = "{foo:{spec}}";
        let parse_ast = parse_fstring(source).unwrap();

        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_parse_fstring_not_nested_spec() {
        let source = "{foo:spec}";
        let parse_ast = parse_fstring(source).unwrap();

        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_parse_empty_fstring() {
        insta::assert_debug_snapshot!(parse_fstring("").unwrap());
    }

    #[test]
    fn test_fstring_parse_selfdocumenting_base() {
        let src = "{user=}";
        let parse_ast = parse_fstring(src).unwrap();

        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_fstring_parse_selfdocumenting_base_more() {
        let src = "mix {user=} with text and {second=}";
        let parse_ast = parse_fstring(src).unwrap();

        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_fstring_parse_selfdocumenting_format() {
        let src = "{user=:>10}";
        let parse_ast = parse_fstring(src).unwrap();

        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_parse_invalid_fstring() {
        assert_eq!(parse_fstring("{5!a"), Err(ExpectedRbrace));
        assert_eq!(parse_fstring("{5!a1}"), Err(ExpectedRbrace));
        assert_eq!(parse_fstring("{5!"), Err(ExpectedRbrace));
        assert_eq!(parse_fstring("abc{!a 'cat'}"), Err(EmptyExpression));
        assert_eq!(parse_fstring("{!a"), Err(EmptyExpression));
        assert_eq!(parse_fstring("{ !a}"), Err(EmptyExpression));

        assert_eq!(parse_fstring("{5!}"), Err(InvalidConversionFlag));
        assert_eq!(parse_fstring("{5!x}"), Err(InvalidConversionFlag));

        assert_eq!(parse_fstring("{a:{a:{b}}}"), Err(ExpressionNestedTooDeeply));

        assert_eq!(parse_fstring("{a:b}}"), Err(UnopenedRbrace));
        assert_eq!(parse_fstring("}"), Err(UnopenedRbrace));
        assert_eq!(parse_fstring("{a:{b}"), Err(UnclosedLbrace));
        assert_eq!(parse_fstring("{"), Err(UnclosedLbrace));

        assert_eq!(parse_fstring("{}"), Err(EmptyExpression));

        // TODO: check for InvalidExpression enum?
        assert!(parse_fstring("{class}").is_err());
    }

    #[test]
    fn test_parse_fstring_not_equals() {
        let source = "{1 != 2}";
        let parse_ast = parse_fstring(source).unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_parse_fstring_equals() {
        let source = "{42 == 42}";
        let parse_ast = parse_fstring(source).unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_parse_fstring_selfdoc_prec_space() {
        let source = "{x   =}";
        let parse_ast = parse_fstring(source).unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_parse_fstring_selfdoc_trailing_space() {
        let source = "{x=   }";
        let parse_ast = parse_fstring(source).unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_parse_fstring_yield_expr() {
        let source = "{yield}";
        let parse_ast = parse_fstring(source).unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }
}
