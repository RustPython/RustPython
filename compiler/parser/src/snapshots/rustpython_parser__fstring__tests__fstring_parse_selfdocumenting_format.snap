---
source: compiler/parser/src/fstring.rs
expression: parse_ast
---
[
    Located {
        location: Location {
            row: 0,
            column: 0,
        },
        custom: (),
        node: Constant {
            value: Str(
                "user=",
            ),
            kind: None,
        },
    },
    Located {
        location: Location {
            row: 0,
            column: 0,
        },
        custom: (),
        node: Constant {
            value: Str(
                "",
            ),
            kind: None,
        },
    },
    Located {
        location: Location {
            row: 0,
            column: 0,
        },
        custom: (),
        node: FormattedValue {
            value: Located {
                location: Location {
                    row: 1,
                    column: 2,
                },
                custom: (),
                node: Name {
                    id: "user",
                    ctx: Load,
                },
            },
            conversion: 0,
            format_spec: Some(
                Located {
                    location: Location {
                        row: 0,
                        column: 0,
                    },
                    custom: (),
                    node: JoinedStr {
                        values: [
                            Located {
                                location: Location {
                                    row: 0,
                                    column: 0,
                                },
                                custom: (),
                                node: Constant {
                                    value: Str(
                                        ">10",
                                    ),
                                    kind: None,
                                },
                            },
                        ],
                    },
                },
            ),
        },
    },
]
